zrml/neo-swaps/src/types/liquidity_tree.rs

```diff
@@ -0,0 +1,1233 @@
+
+    /// Return the maximum allowed depth of the tree.
+    fn max_depth() -> u32;
+
+    /// Return the maximum allowed amount of nodes in the tree.
+    fn max_node_count() -> u32;
+}
+
+impl<T, U> LiquidityTreeHelper<T> for LiquidityTree<T, U>
+where
+    T: Config,
+    U: Get<u32>,
+{
+    fn propagate_fees_to_node(&mut self, index: u32) -> DispatchResult {
+        let path = self.path_to_node(index)?;
+        for i in path {
+            self.propagate_fees(i)?;
+        }
+        Ok(())
+    }
+
+    fn propagate_fees(&mut self, index: u32) -> DispatchResult {
+        let node = self.get_node(index)?;
+        if node.total_stake()? == Zero::zero() {
+            return Ok(()); // Don't propagate if there are no LPs under this node.
+        }
+        if node.is_weak_leaf() {
+            self.mutate_node(index, |node| {
+                node.fees = node.fees.checked_add_res(&node.lazy_fees)?;
+                Ok(())
+            })?;
+        } else {
+            // Temporary storage to ensure that the borrow checker doesn't get upset.
+            let node_descendant_stake = node.descendant_stake;
+            // The lazy fees that will be propagated down the tree.
+            let mut remaining_lazy_fees =
+                node.descendant_stake.bmul_bdiv(node.lazy_fees, node.total_stake()?)?;
+            // The fees that stay at this node.
+            let fees = node.lazy_fees.checked_sub_res(&remaining_lazy_fees)?;
+            self.mutate_node(index, |node| {
+                node.fees = node.fees.checked_add_res(&fees)?;
+                Ok(())
+            })?;
+            let (opt_lhs_index, opt_rhs_index) = self.children(index)?.into();
+            if let Some(lhs_index) = opt_lhs_index {
+                self.mutate_node(lhs_index, |lhs_node| {
+                    // The descendant's share of the stake:
+                    let child_lazy_fees = lhs_node
+                        .total_stake()?
+                        .bmul_bdiv(remaining_lazy_fees, node_descendant_stake)?;
+                    lhs_node.lazy_fees = lhs_node.lazy_fees.checked_add_res(&child_lazy_fees)?;
+                    remaining_lazy_fees = remaining_lazy_fees.checked_sub_res(&child_lazy_fees)?;
+                    Ok(())
+                })?;
+            }
+            if let Some(rhs_index) = opt_rhs_index {
+                self.mutate_node(rhs_index, |rhs_node| {
+                    rhs_node.lazy_fees =
+                        rhs_node.lazy_fees.checked_add_res(&remaining_lazy_fees)?;
+                    Ok(())
+                })?;
+            }
+        }
+        self.mutate_node(index, |node| {
+            node.lazy_fees = Zero::zero();
+            Ok(())
+        })?;
+        Ok(())
+    }
+
+    fn children(&self, index: u32) -> Result<LiquidityTreeChildIndices, DispatchError> {
+        let calculate_child =
+            |child_index: u32| Some(child_index).filter(|&i| i < self.node_count());
```

Comment:
The idea here is that if the maximum number of nodes is not yet reached, then we return `None` here to signal that one of the children is in an empty spot.

Reply:
So assuming that other parts of the logic always ensure that `self.node_count() <= self.max_node_count()`, this function returns `None` either if a child node would occupy a new empty node or that a child would occupy a node that is impossible to occupy (because it would exceed the max number of nodes). Is that the excepted behavior?

Reply:
Yes, that's correct. I've updated the documentation: 33c4ec7db9b0ad7fe80aceaae0188cf885894016